SU2_CFD/include/numerics/NEMO/CNEMONumerics.hpp

```diff
@@ -155,20 +154,293 @@ class CNEMONumerics : public CNumerics {
    * \param[out] val_Proj_Jac_Tensor_j - Pointer to the projected viscous Jacobian at point j.
    */
-  void GetViscousProjJacs(su2double *val_Mean_PrimVar,
-                          su2double **val_Mean_GradPrimVar,
-                          su2double *val_Mean_Eve,
-                          su2double *val_Mean_Cvve,
-                          su2double *val_diffusion_coeff,
-                          su2double val_laminar_viscosity,
-                          su2double val_eddy_viscosity,
-                          su2double val_thermal_conductivity,
+  void GetViscousProjJacs(const su2double *val_Mean_PrimVar,
+                          su2double *val_Mean_Eve, const su2double *val_Mean_Cvve,
+                          const su2double *val_diffusion_coeff, su2double val_laminar_viscosity,
+                          su2double val_eddy_viscosity, su2double val_thermal_conductivity,
                           su2double val_thermal_conductivity_ve,
-                          su2double val_dist_ij, su2double *val_normal,
-                          su2double val_dS, su2double *val_Fv,
+                          su2double val_dist_ij, 
+                          const su2double *val_normal,
+                          su2double val_dS, const su2double *val_Fv,
                           su2double **val_Jac_i, su2double **val_Jac_j,
                           const CConfig *config);
 
+  /*!
+   * \brief TSL-Approximation of Viscous NS Jacobians for arbitrary equations of state.
+   */
+  template <unsigned short NVAR, unsigned short NSPECIES>
+  void ComputeViscousJacs_impl(const su2double *val_Mean_PrimVar,
+                               su2double *val_Mean_Eve,
+                               const su2double *val_Mean_Cvve,
+                               const su2double *val_diffusion_coeff,
+                               su2double val_laminar_viscosity,
+                               su2double val_eddy_viscosity,
+                               su2double val_thermal_conductivity,
+                               su2double val_thermal_conductivity_ve,
+                               su2double val_dist_ij, const su2double *val_normal,
+                               su2double val_dS, const su2double *val_Fv,
+                               su2double **val_Jac_i, su2double **val_Jac_j,
+                               const CConfig *config){
+
+    /*--- Initialize matrices ---*/
+    Matrix<NVAR, NVAR> dFdVi, dFdVj, dVdUi, dVdUj;
+    /*--- Need to avoid size of 1 otherwise the type becomes a vector (without [i][j]). ---*/
+    constexpr unsigned short nSpeciesDummy = (NSPECIES == 1) ? 2 : NSPECIES;
+    Matrix<nSpeciesDummy, nSpeciesDummy> dJdr_i, dJdr_j;
+
+    /*--- Play tricks on the compiler, in static mode use NVAR from the template, in dynamic mode
+          use nVar from the class or from the arguments (if you make it static it needs to be an argument). ---*/
+    const auto nVar = NVAR;
+    const auto nSpecies = NSPECIES;
```

Comment:
You have to play the actual tricks :) not sure if you have nSpecies as a member variable, otherwise you can also get it from the config.
```suggestion
    const auto nVar = (NVAR != DynamicSize) ? NVAR : this->nVar;
    const auto nSpecies = (NSPECIES != DynamicSize) ? NSPECIES : this->nSpecies;
```

Reply:
True!  Im no good these tricks.